File: src/scorer/election-assessors/benefic-patterns.ts
import { RASI_NAMES } from '../../data/rasis';
import type { ElectionChartSnapshot } from '../../lib/election-chart-api';
import {
  evaluateFullAspect,
} from './chart-geometry';
import type { PrimitiveOutcome } from './contracts';
import { completePlanetPositions } from './event-admission';
import {
  classifyNaturalGrahaNatures,
  evaluateExistentialBeneficHouseSet,
} from './graha-nature';

export const MALE_RASIS = new Set(RASI_NAMES.filter((_, index) => index % 2 === 0));
export const KENDRA_HOUSES = [1, 4, 7, 10] as const;
const CLASSICAL_GRAHAS = [
  'Surya', 'Chandra', 'Kuja', 'Budha', 'Guru', 'Shukra', 'Shani',
] as const;
const ALL_GRAHAS = new Set([
  ...CLASSICAL_GRAHAS, 'Rahu', 'Ketu',
]);
const RASI_ROUNDING_GUARD_DEGREES = 0.01;

function prefixed(label: string, outcome: PrimitiveOutcome): string[] {
  return outcome.evidence.map(item => `${label}: ${item}`);
}

function maleRasiBeneficAspect(chart: ElectionChartSnapshot): PrimitiveOutcome {
  const positions = completePlanetPositions(chart, ALL_GRAHAS);
  const nature = classifyNaturalGrahaNatures(chart);
  if (!positions || !nature.complete) {
    return {
      status: 'unknown',
      evidence: nature.evidence.length
        ? nature.evidence
        : ['Complete canonical nine-graha facts are unavailable.'],
    };
  }

  const witnesses: string[] = [];
  let unresolved = false;
  for (const targetName of CLASSICAL_GRAHAS) {
    const target = positions.get(targetName)!;
    const targetNature = nature.natures[targetName];
    const nearBoundary = Math.min(target.degree, 30 - target.degree)
      <= RASI_ROUNDING_GUARD_DEGREES;
    if (!MALE_RASIS.has(target.rashi)) {
      unresolved ||= nearBoundary && ['benefic', 'unknown'].includes(targetNature);
      continue;
    }

    const beneficSources = CLASSICAL_GRAHAS.filter(name =>
      name !== targetName && nature.natures[name] === 'benefic');
    const unresolvedSources = CLASSICAL_GRAHAS.filter(name =>
      name !== targetName && nature.natures[name] === 'unknown');
    const beneficAspect = evaluateFullAspect({
      kind: 'planet_receives_full_aspect',
      planet: targetName,
      aspectors: beneficSources,
    }, positions);
    const unresolvedAspect = evaluateFullAspect({
      kind: 'planet_receives_full_aspect',
      planet: targetName,
      aspectors: unresolvedSources,
    }, positions);
    if (targetNature === 'benefic' && beneficAspect.status === 'pass') {
      const aspectors = beneficAspect.evidence[0].split(': ', 2)[1].replace(/\.$/, '');
      witnesses.push(
        `${targetName} in ${target.rashi} receives full Graha Drishti from ${aspectors}`,
      );
    } else if (
      (targetNature === 'unknown' && (
        beneficAspect.status !== 'fail' || unresolvedAspect.status !== 'fail'
      ))
      || (targetNature === 'benefic' && (
        beneficAspect.status === 'unknown' || unresolvedAspect.status !== 'fail'
      ))
    ) {
      unresolved = true;
    }
  }

  if (witnesses.length) {
    return {
      status: 'pass',
      evidence: [`Resolved natural-benefic witness: ${witnesses.join('; ')}.`],
    };
  }
  if (unresolved) {
    return {
      status: 'unknown',
      evidence: ['A possible natural-benefic male-Rasi aspect witness is unresolved.'],
    };
  }
  return {
    status: 'fail',
    evidence: [
      'No resolved natural benefic in an odd Rasi receives a full classical aspect from another resolved natural benefic.',
    ],
  };
}

export function evaluateBeneficKendraOrMaleRasiAspect(
  chart: ElectionChartSnapshot,
  options: { houseFrameUncertain?: boolean } = {},
): PrimitiveOutcome {
  if (
    !options || typeof options !== 'object' || Array.isArray(options)
    || (options.houseFrameUncertain !== undefined
      && typeof options.houseFrameUncertain !== 'boolean')
  ) {
    return { status: 'unknown', evidence: ['The benefic-pattern configuration is malformed.'] };
  }
  if (options.houseFrameUncertain) {
    return {
      status: 'unknown',
      evidence: ['The validated local-Lagna house frame is unavailable or conflicting.'],
    };
  }
  if (!chart || typeof chart !== 'object' || Array.isArray(chart)) {
    return {
      status: 'unknown',
      evidence: ['Complete canonical nine-graha facts are unavailable.'],
    };
  }

  const kendra = evaluateExistentialBeneficHouseSet(chart, KENDRA_HOUSES);
  const maleRasiAspect = maleRasiBeneficAspect(chart);
  const evidence = [
    ...prefixed('Kendra arm', kendra),
    ...prefixed('Male-Rasi aspect arm', maleRasiAspect),
  ];
  if ([kendra.status, maleRasiAspect.status].includes('pass')) {
    return { status: 'pass', evidence };
  }
  if ([kendra.status, maleRasiAspect.status].includes('unknown')) {
    return { status: 'unknown', evidence };
  }
  return { status: 'fail', evidence };
}

File: telugu_panchangam/personal/election_assessors/benefic_patterns.py
"""Source-neutral natural-benefic placement and aspect patterns."""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from ...panchangam_names import RASHI_NAMES
from .chart_geometry import (
    FULL_ASPECT_OFFSETS,
    RASHI_ROUNDING_GUARD_DEGREES,
    evaluate_full_aspect,
)
from .contracts import PrimitiveOutcome
from .event_admission import planet_positions
from .graha_nature import (
    classify_natural_graha_natures,
    evaluate_existential_benefic_house_set,
)

MALE_RASIS = frozenset(RASHI_NAMES[::2])
KENDRA_HOUSES = (1, 4, 7, 10)
_CLASSICAL_GRAHAS = tuple(FULL_ASPECT_OFFSETS)


def _prefixed(label: str, outcome: PrimitiveOutcome) -> tuple[str, ...]:
    return tuple(f'{label}: {item}' for item in outcome.evidence)


def _male_rasi_benefic_aspect(
    chart: Mapping[str, Any],
) -> PrimitiveOutcome:
    positions = planet_positions(chart)
    nature = classify_natural_graha_natures(chart)
    if positions is None or not nature.complete:
        return PrimitiveOutcome('unknown', nature.evidence)

    witnesses: list[str] = []
    unresolved = False
    for target_name in _CLASSICAL_GRAHAS:
        target = positions[target_name]
        target_nature = nature.natures[target_name]
        near_boundary = min(target.degree, 30 - target.degree) <= (
            RASHI_ROUNDING_GUARD_DEGREES
        )
        if target.rashi not in MALE_RASIS:
            unresolved = unresolved or (
                near_boundary and target_nature in {'benefic', 'unknown'}
            )
            continue

        benefic_sources = [
            name
            for name in _CLASSICAL_GRAHAS
            if name != target_name
            and nature.natures[name] == 'benefic'
        ]
        unresolved_sources = [
            name
            for name in _CLASSICAL_GRAHAS
            if name != target_name and nature.natures[name] == 'unknown'
        ]
        benefic_aspect = evaluate_full_aspect(
            {'planet': target_name, 'aspectors': benefic_sources}, positions
        )
        unresolved_aspect = evaluate_full_aspect(
            {'planet': target_name, 'aspectors': unresolved_sources}, positions
        )
        if target_nature == 'benefic' and benefic_aspect.status == 'pass':
            aspectors = benefic_aspect.evidence[0].split(': ', 1)[1].removesuffix('.')
            witnesses.append(
                f'{target_name} in {target.rashi} receives full Graha Drishti '
                f'from {aspectors}'
            )
        elif (
            target_nature == 'unknown'
            and (
                benefic_aspect.status != 'fail'
                or unresolved_aspect.status != 'fail'
            )
        ) or (
            target_nature == 'benefic'
            and (
                benefic_aspect.status == 'unknown'
                or unresolved_aspect.status != 'fail'
            )
        ):
            unresolved = True

    if witnesses:
        return PrimitiveOutcome(
            'pass',
            ('Resolved natural-benefic witness: ' + '; '.join(witnesses) + '.',),
        )
    if unresolved:
        return PrimitiveOutcome(
            'unknown',
            ('A possible natural-benefic male-Rasi aspect witness is unresolved.',),
        )
    return PrimitiveOutcome(
        'fail',
        (
            'No resolved natural benefic in an odd Rasi receives a full '
            + 'classical aspect from another resolved natural benefic.',
        ),
    )


def evaluate_benefic_kendra_or_male_rasi_aspect(
    chart: Mapping[str, Any],
    *,
    house_frame_uncertain: bool = False,
) -> PrimitiveOutcome:
    """Evaluate the disclosed two-arm benefic pattern without an event effect."""
    if type(house_frame_uncertain) is not bool:
        return PrimitiveOutcome(
            'unknown', ('The benefic-pattern configuration is malformed.',)
        )
    if house_frame_uncertain:
        return PrimitiveOutcome(
            'unknown',
            ('The validated local-Lagna house frame is unavailable or conflicting.',),
        )
    if not isinstance(chart, Mapping):
        return PrimitiveOutcome(
            'unknown', ('Complete canonical nine-graha facts are unavailable.',)
        )

    kendra = evaluate_existential_benefic_house_set(chart, KENDRA_HOUSES)
    male_rasi_aspect = _male_rasi_benefic_aspect(chart)
    evidence = (
        *_prefixed('Kendra arm', kendra),
        *_prefixed('Male-Rasi aspect arm', male_rasi_aspect),
    )
    if 'pass' in {kendra.status, male_rasi_aspect.status}:
        return PrimitiveOutcome('pass', evidence)
    if 'unknown' in {kendra.status, male_rasi_aspect.status}:
        return PrimitiveOutcome('unknown', evidence)
    return PrimitiveOutcome('fail', evidence)
